tests/MongoDB.Driver.Core.Tests/Core/Authentication/AvsSignatureVersion4Tests.cs

```diff
@@ -0,0 +1,130 @@
+﻿/* Copyright 2020-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Net;
+using System.Security;
+using System.Text;
+using FluentAssertions;
+using MongoDB.Driver.Core.Authentication;
+using MongoDB.Driver.Core.Authentication.Sspi;
+using MongoDB.Driver.Core.Connections;
+using MongoDB.Driver.Core.Misc;
+using Xunit;
+
+namespace MongoDB.Driver.Core.Tests.Core.Authentication
+{
+    public class AvsSignatureVersion4Tests
+    {
+        [Theory]
+        [InlineData("sts.amazonaws.com", "us-east-1")]
+        [InlineData("first", "us-east-1")]
+        [InlineData("first.second", "second")]
+        [InlineData("first.second.third", "second")]
+        public void GetRegionShouldWorkAsExpected(string host, string expectedRegion)
+        {
+            var region = AwsSignatureVersion4.GetRegion(host);
+
+            region.Should().Be(expectedRegion);
+        }
+
+        [Fact]
+        public void GetCanonicalHeadersShouldWorkAsExpected()
+        {
+            var timestamp = new DateTime(2020, 03, 12, 14, 23, 46).ToString("yyyyMMddTHHmmssZ");
+            var requestHeaders = new SortedDictionary<string, string>();
+            requestHeaders["X-MongoDB-GS2-CB-Flag"] = "n";
+            requestHeaders["Content-Type"] = "application/x-www-form-urlencoded";
+            requestHeaders["X-Amz-Date"] = timestamp;
+            requestHeaders["Content-Length"] = "42";
```

Comment:
This should be based on the length of the body ("Action=GetCallerIdentity&Version=2011-06-15") which should be 43 characters long.